recognition/scoring.py

```diff
@@ -664,3 +665,52 @@ def calc_wer(self):
         os.chdir(old_dir)
 
         return wer
+
+
+class IntersectStmCtm(Job):
+    def __init__(self, stm_path: tk.Path, ctm_path: tk.Path):
+        self.stm_path = stm_path
+        self.ctm_path = ctm_path
+
+        self.out_stm = self.output_path("intersected.stm")
+        self.out_ctm = self.output_path("intersected.ctm")
+
+        self.rqmt = None
+
+    def tasks(self):
+        yield Task("run", resume="run", mini_task=self.rqmt is None, rqmt=self.rqmt)
+
+    def run(self):
+        stm_lines = collections.defaultdict(list)
+        ctm_lines = collections.defaultdict(list)
+
+        with uopen(self.stm_path, "rt") as stm_in:
+            for line in stm_in:
+                if line.startswith(";"):
+                    stm_lines[None].append(line)
+                else:
+                    recording = line.split()[0]
+                    stm_lines[recording].append(line)
+
+        segment_comments = []
+        with uopen(self.ctm_path, "rt") as ctm_in:
+            for line_num, line in enumerate(ctm_in, 1):
+                if line_num == 1 and line.startswith(";; <name>"):  # header
+                    ctm_lines[None].append(line)
+                elif line.startswith(";"):
+                    segment_comments.append(line)
+                else:
+                    recording = line.split()[0]
+                    if segment_comments:
+                        ctm_lines[recording].extend(segment_comments)
+                        segment_comments = []
+                    ctm_lines[recording].append(line)
+
+        common_recordings = set(stm_lines.keys()).intersection(set(ctm_lines.keys()))
+
+        for path, lines in [(self.out_stm, stm_lines), (self.out_ctm, ctm_lines)]:
+            with uopen(path, "wt") as out:
+                out.write("".join(lines[None]))
+                del lines[None]
```

Comment:
why `del lines[None]`? I assume because `None` would be in the set `common_recordings` and in the for-loop below we would also iterate again over it and write the header-lines a second time.

But now we still do iterate over it and when `None` shows up, the defaultdict nature of `lines` prevents a crash and leads to `out.write("".join([])` which if fine I guess.
Maybe semantically more nice would be to remove `None` from the set of `common_recordings` e.g.
```
common_recordings = set(stm_lines.keys()).intersection(set(ctm_lines.keys())) - {None}
```

I leave this open.